pkg/services/sqlstore/user.go

```diff
@@ -192,6 +245,12 @@ func (ss *SQLStore) GetUserByLogin(ctx context.Context, query *models.GetUserByL
 			return models.ErrUserNotFound
 		}
 
+		if ss.Cfg.CaseInsensitiveLogin {
+			if err := ss.userCaseInsensitiveLoginConflict(ctx, sess, user.Login, user.Email); err != nil {
```

Comment:
thought: again, not sure how relevant it is to this PR, but are we ok with the case where one user's login is another user's email and how that would impact the `LoginOrEmail` retrieval flow? I'm not sure whether user creation/update flow allows for such cases to arise, but it's something that currently wouldn't be flagged up by `userCaseInsensitiveLoginConflict`, so we should double-check it.

Reply:
it's a fair question, and likely something that wasn't considered when we started using email addresses as login names.  We need to consider how to handle the case where 2 users have email addresses or logins that only differ in case, so probably worth considering this as well.  The safest approach may be to clarify the resolution rules, eg the way that GetUserByLogin would always return the user with the matching login.  Similarly, for case insensitive matching it may make sense to either always return the oldest matching account, or to prefer an exact match over a case-insensitive match.  That gets tricky when combined with login vs email matching though...

Here again I'm unsure that we actually want to error out here since it'll block anyone from fixing the problem if they can't log in and/or can't load the affected user to update them.

Reply:
Summarizing the replies to the 2 other threads:

- Login or Email conflict

I think we should tackle this in a separate issue as it already affects grafana out of the configuration option. So the fix/clarification of the behavior should cover both scenarios.

- Case sensitive conflicts: 

Blocking the users until the conflict is solved stops us from accidentally merging unrelated users and puts the onus on the server admins. 

We intend to give server admins better tools for this in the next PR.

If we don't force users to fix the login/email conflicts, we won't be able to add case insensitive DB checks to prevent re-adding case sensitive behavior by accident.

If this blocks instances for now, they would only need to disable the config option (which is disabled by default for now)

https://docs.google.com/presentation/d/1DBoES_ZzB2OElF2qK0LUz7p2Um_TSbe-3IG9BB5Kajs/edit#slide=id.g133a3c2633b_1_37